daemon.ContainerLogs(): fix resource leak on follow

kolyshkin opened the pull request:
When `daemon.ContainerLogs()` is called with `options.follow=true`
(as in `docker logs --follow`), the `loggerutils.followLogs()`
function never returns (even then the logs consumer is gone).
As a result, all the resources associated with it (including
an opened file descriptor for the log file being read, two FDs
for a pipe, and two FDs for inotify watch) are never released.

If this is repeated (such as by running `docker logs --follow`
and pressing `Ctrl-C` a few times), this results in DoS caused by
either hitting the limit of inotify watches, or the limit of
opened files. The only cure is daemon restart.

Apparently, what happens is:

1. logs consumer is gone, properly calling `(*LogWatcher).Close()`

2. `WatchClose()` is properly handled by a dedicated goroutine in
`followLogs()`, cancelling the context.

3. Upon receiving the `ctx.Done()`, the code tries to _synchronously_
send the log message to the `logWatcher.Msg` channel. Since the
channel reader is gone, this operation blocks forever, and
resource cleanup set up in defer statements at the beginning
of `followLogs()` never happens.

The fix is to remove the synchronous send.

This commit also removes the code after it, which tried to
read and send the rest of the log file. The code in question
first appeared in commit c0391bf ("Split reader interface from
logger interface"), but it is still unclear why it is/was
needed (it makes no sense to write logs once the consumer
signaled it is no longer interested in those).

There are a few alternative approaches to fix the issue,
such as:

a. amend the consumer to read all the data from the channel
until it is closed. This can be done in, say, `(*LogWatcher) Close()`.
For one thing, this looks ugly (why read and write useless data?).
Also, this approach requires closing the channel from the sending
side, which, given the amount of exit paths, looks less elegant
than a solution in this commit.

b. always write to the channel asynchronously (i.e. in a non-blocking
fashion, i.e. inside a select). This is actually what happens
in the code after this commit.

Should fix https://github.com/moby/moby/issues/37391

![Щас спою!](https://img-fotki.yandex.ru/get/62627/12959330.fb/0_147c37_f66a40bf_orig)
*image from https://dubikvit.livejournal.com/409094.html*

kolyshkin committed:
daemon.ContainerLogs: minor debug logging cleanup

This code has many return statements, for some of them the
"end logs" or "end stream" message was not printed, giving
the impression that this "for" loop never ended.

Make sure that "begin logs" is to be followed by "end logs".

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon.ContainerLogs(): fix resource leak on follow

When daemon.ContainerLogs() is called with options.follow=true
(as in "docker logs --follow"), the "loggerutils.followLogs()"
function never returns (even then the logs consumer is gone).
As a result, all the resources associated with it (including
an opened file descriptor for the log file being read, two FDs
for a pipe, and two FDs for inotify watch) are never released.

If this is repeated (such as by running "docker logs --follow"
and pressing Ctrl-C a few times), this results in DoS caused by
either hitting the limit of inotify watches, or the limit of
opened files. The only cure is daemon restart.

Apparently, what happens is:

1. logs producer (a container) is gone, calling (*LogWatcher).Close()
for all its readers (daemon/logger/jsonfilelog/jsonfilelog.go:175).

2. WatchClose() is properly handled by a dedicated goroutine in
followLogs(), cancelling the context.

3. Upon receiving the ctx.Done(), the code in followLogs()
(daemon/logger/loggerutils/logfile.go#L626-L638) keeps to
send messages _synchronously_ (which is OK for now).

4. Logs consumer is gone (Ctrl-C is pressed on a terminal running
"docker logs --follow"). Method (*LogWatcher).Close() is properly
called (see daemon/logs.go:114). Since it was called before and
due to to once.Do(), nothing happens (which is kinda good, as
otherwise it will panic on closing a closed channel).

5. A goroutine (see item 3 above) keeps sending log messages
synchronously to the logWatcher.Msg channel. Since the
channel reader is gone, the channel send operation blocks forever,
and resource cleanup set up in defer statements at the beginning
of followLogs() never happens.

Alas, the fix is somewhat complicated:

1. Distinguish between close from logs producer and logs consumer.
To that effect,
 - yet another channel is added to LogWatcher();
 - {Watch,}Close() are renamed to {Watch,}ProducerGone();
 - {Watch,}ConsumerGone() are added;

*NOTE* that ProducerGone()/WatchProducerGone() pair is ONLY needed
in order to stop ConsumerLogs(follow=true) when a container is stopped;
otherwise we're not interested in it. In other words, we're only
using it in followLogs().

2. Code that was doing (logWatcher*).Close() is modified to either call
ProducerGone() or ConsumerGone(), depending on the context.

3. Code that was waiting for WatchClose() is modified to wait for
either ConsumerGone() or ProducerGone(), or both, depending on the
context.

4. followLogs() are modified accordingly:
 - context cancellation is happening on WatchProducerGone(),
and once it's received the FileWatcher is closed and waitRead()
returns errDone on EOF (i.e. log rotation handling logic is disabled);
 - due to this, code that was writing synchronously to logWatcher.Msg
can be and is removed as the code above it handles this case;
 - function returns once ConsumerGone is received, freeing all the
resources -- this is the bugfix itself.

While at it,

1. Let's also remove the ctx usage to simplify the code a bit.
It was introduced by commit a69a59ffc7e3d ("Decouple removing the
fileWatcher from reading") in order to fix a bug. The bug was actually
a deadlock in fsnotify, and the fix was just a workaround. Since then
the fsnofify bug has been fixed, and a new fsnotify was vendored in.
For more details, please see
https://github.com/moby/moby/pull/27782#issuecomment-416794490

2. Since `(*filePoller).Close()` is fixed to remove all the files
being watched, there is no need to explicitly call
fileWatcher.Remove(name) anymore, so get rid of the extra code.

Should fix https://github.com/moby/moby/issues/37391

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented:
Looks like there's a failure; https://jenkins.dockerproject.org/job/Docker-PRs/50259/console 

That test was marked as flaky a couple of times; https://github.com/moby/moby/search?q=TestLogsFollowSlowStdoutConsumer&type=Issues could that flakiness be related?

```
05:52:22 FAIL: docker_cli_logs_test.go:214: DockerSuite.TestLogsFollowSlowStdoutConsumer
05:52:22 
05:52:22 docker_cli_logs_test.go:246:
05:52:22     c.Assert(actual, checker.Equals, expected)
05:52:22 ... obtained int = 53412
05:52:22 ... expected int = 150000
```

kolyshkin commented:
> 05:52:22 FAIL: docker_cli_logs_test.go:214: DockerSuite.TestLogsFollowSlowStdoutConsumer

Fortunately I know a bit about this test. OK, so now we know why this code was needed :) Need to think.

cpuguy83 commented:
`logWatcher.Close()` is called when the client is gone:
https://github.com/moby/moby/blob/e1584514c15f6ad91b9ef5bc530a1eef46f18267/daemon/logs.go#L106-L114

Where `cLog` is the created log driver in cases where the container was already stopped and `logs` is the log watcher.

So we should not be trying to send logs when the log watcher is closed.
What may need to happen is to make sure to drain the message chan in `daemon/logs.go` after the log watcher is closed, otherwise we can end up in an easy to trigger race where a message send is happening while we are closing.
The buffer of this channel is also insanely high... a number which was just picked because someone asked for it.... I don't think we need a buffer, or if we do have a buffer it can be very small.

It would also be nice perhaps to change the log reader interface to take a context instead of closing the log watcher, may help simplify some of the logic.

Honestly, there are too many goroutines here (one layer added *relatively* recently to support swarm logs).

kolyshkin commented:
The `(*logWatcher).Close()` is also called from here:
https://github.com/moby/moby/blob/09f5e9dae3b83ee9eaa09b8097134d5d4251fce8/daemon/logger/jsonfilelog/jsonfilelog.go#L169-L177

So, it is called twice, once upon container stop, once upon consumer is gone. Whoever calls it first depends on the situation. In this scenario the logger is fast and the consumer is slow(er), so the first call is from container stop, thus (with the commit in this PR) the log consumer does not receive the remainder of the log.

If we comment out this line:
https://github.com/moby/moby/blob/09f5e9dae3b83ee9eaa09b8097134d5d4251fce8/daemon/logger/jsonfilelog/jsonfilelog.go#L175

`docker logs --follow` works fine (with the commit from this PR), but it never exits (which, to my mind, is the correct behavior as the container could be started again, but is different from the current behavior).

For now, it seems like the best behavior would be to distinguish between "container stopped logging" and "logs consumer is gone" events, and act accordingly, i.e:

1. when "logs consumer is gone" received, stop immediately. In order to not have a race, the receiver should be inside the same select as the message sender (I guess).
2. when "container stopped logging" is received, disable the log rotation handling, otherwise continue until EOF.

Unfortunately it further complicates the code, which I wanted to avoid :-\

kolyshkin commented:
> It would also be nice perhaps to change the log reader interface to take a context instead of closing the log watcher, may help simplify some of the logic.

We have to react to two distinct events here: 

1. logs producer (a container) gone. In this case `followLogs()` should stop watching for and reacting to logrotate events, and treat EOF as being done, otherwise keep sucking and pushing logs.
2. logs consumer (daemon.ContainerLogs()) gone. In this case we should clean up and be done.

A single context won't be enough here. Sigh :(

kolyshkin commented:
OK, a test case `DockerSuite.TestLogsFollowGoroutinesNoOutput` uncovered a problem that I have fixed.

TODO: write a test case for https://github.com/moby/moby/issues/37391

kolyshkin commented:
ppc failure on DockerSwarmSuite.TearDownTest is https://github.com/moby/moby/issues/33041

kolyshkin commented:
OK, I removed the fixing commit and added the test case, doing a CI cycle to see if the newly added test case is detecting the issue.

cpuguy83 commented:
Yeah, so basically it looks like we just need to split notifications up. Would be nice not to modify the log watcher for this.

kolyshkin commented:
> Yeah, so basically it looks like we just need to split notifications up.
> Would be nice not to modify the log watcher for this.

Here's my current solution (will push here once test case is checked to work): https://github.com/kolyshkin/moby/commit/16bc8991a

It is somewhat ugly and big, so I am wide open to any ideas how to make it look better.

kolyshkin commented:
Test case fixes and should be good to go. Waiting for CI to fail!

kolyshkin commented:
Failing as expected :grinning: 

on z:
```
00:18:43.093 --- FAIL: TestLogsFollowGoroutineLeak (12.36s)
00:18:43.093 	logs_test.go:110: daemon resources before: goroutines: 77, file descriptors: 59
00:18:43.093 	logs_test.go:131: exit after reading 2546100 bytes
00:18:43.093 	logs_test.go:90: daemon resources after: goroutines: 79, file descriptors: 64
00:18:43.093 	logs_test.go:167: Leaked goroutines: 2, file descriptors: 5
```

on experimental:
```
00:24:42.959 --- FAIL: TestLogsFollowGoroutineLeak (11.97s)
00:24:42.959 	logs_test.go:110: daemon resources before: goroutines: 77, file descriptors: 59
00:24:42.959 	logs_test.go:131: exit after reading 1005220 bytes
00:24:42.959 	logs_test.go:90: daemon resources after: goroutines: 79, file descriptors: 64
00:24:42.959 	logs_test.go:167: Leaked goroutines: 2, file descriptors: 5
```

on janky:
```
00:38:54.787 --- FAIL: TestLogsFollowGoroutineLeak (12.20s)
00:38:54.787 	logs_test.go:110: daemon resources before: goroutines: 77, file descriptors: 59
00:38:54.788 	logs_test.go:131: exit after reading 780560 bytes
00:38:54.788 	logs_test.go:90: daemon resources after: goroutines: 79, file descriptors: 64
00:38:54.788 	logs_test.go:167: Leaked goroutines: 2, file descriptors: 5
```

on ppc:
```
00:43:28.488 --- FAIL: TestLogsFollowGoroutineLeak (14.16s)
00:43:28.488 	logs_test.go:110: daemon resources before: goroutines: 79, file descriptors: 61
00:43:28.488 	logs_test.go:131: exit after reading 897580 bytes
00:43:28.489 	logs_test.go:90: daemon resources after: goroutines: 81, file descriptors: 66
00:43:28.489 	logs_test.go:167: Leaked goroutines: 2, file descriptors: 5
```

Now adding the fix!

kolyshkin commented:
z needs to be restarted

experimental failed on DockerSwarmSuite.TestSwarmLockUnlockCluster, flaky (see https://github.com/moby/moby/issues/34051)

thaJeztah commented:
@kolyshkin this is ready for review?

kolyshkin commented:
> this is ready for review?

Totally :sunglasses:

cpuguy83 commented on daemon/logger/journald/read.go:
```diff
@@ -299,7 +299,11 @@ func (s *journald) followJournal(logWatcher *logger.LogWatcher, j *C.sd_journal,
+	case <-logWatcher.WatchProducerGone():
+		// Notify the other goroutine that its work is done.
+		C.close(pfd[1])
+		cursor = <-newCursor
```
Wouldn't we still want to send the rest of the logs?

kolyshkin replied:
this just mimics the old behavior (and minimizes the patch impact); as I mentioned in commit message:

> 2. Most of the code waiting for WatchClose() is modified to wait for
> either ConsumerGone() or ProducerGone(), which mimics the old behavior,
> except for a few places where it's perfectly clear which one we
> are to wait for. TODO: further specification for some cases might
> be needed to fix other bugs.

so yes, you're probably right but this can be addressed later

thaJeztah commented:
Not sure I fully grasp @cpuguy83's last comment, but IIRC, `logs -f` also allowed you to stay attached when a container is stopped/exited, and continue streaming when it is started again.

Don't know if I remember that correctly, but if so, we need to make sure that behavior doesn't change

kolyshkin commented:
> IIRC, logs -f also allowed you to stay attached when a container is stopped/exited, and continue streaming when it is started again.

I fully agree it _should_ work that way, also I have expected it to work that way. Unfortunately currently it does not (even without commits from this PR). I tested both docker-ce-18.06 and docker-ce-17.03 and they both exhibit the same behavior: on container stop (not remove!) `docker logs -f` exits.

To reproduce:
```
term_1# docker run -d busybox sh -c "while true; do date; sleep 0.1; done"

term_2# docker logs -f CONTAINER_ID

term_1# docker stop CONTAINER_ID
```

on term_2, `docker logs -f` exits.

Filed a separate issue https://github.com/moby/moby/issues/37630 to not put all the eggs in the same basket; let's discuss this aspect in there.

kolyshkin commented:
> So I would like to see this refactored a bit. It's really weird that the log watcher is concerned with the log producer being done. 
> I would expect this to be handled in the logger itself.

Can you please elaborate? This PR did not add log watcher reaction on log producer being done -- it was that way before (`(*LogWatcher).Close()` was called from _both_ producer and consumer when they were done). The only thing this PR does is adding a distinction between the producer and the consumer, and an appropriate reaction.

Please let me know in more details what you have in mind and I'll work on that.

cpuguy83 commented on daemon/logger/adapter.go:
```diff
@@ -127,9 +130,9 @@ func (a *pluginAdapterWithRead) ReadLogs(config ReadConfig) *LogWatcher {
 			case watcher.Msg <- msg:
-			case <-watcher.WatchClose():
-				// make sure the message we consumed is sent
-				watcher.Msg <- msg
```
Shouldn't we still send this in the case of producer gone?

kolyshkin replied:
Well, this is a blocking send, and it will block here forever in case the consumer is gone.

cpuguy83 replied:
```go
select {
case <-watcher.Msg:
case <-watcher.WatchConsumerGone():
}
```

kolyshkin replied:
Yes, this is what I was thinking as well, but in this case the code will look like this:

```go
select {
￼ 	case watcher.Msg <- msg:
	case <-watcher.WatchConsumerGone():
		return
	case <-watcher.WatchProducerGone():
		select {
		case watcher.Msg <- msg:
		case <-watcher.WatchConsumerGone():
		}
		return
}￼
```

Does it really have to look that ugly (in order to work correctly)?

cpuguy83 replied:
Could wrap this in a function... like `send(msg logger.Message, chMsg <-logger.Message, watcher logger.Watcher) <-chan struct`
With the return being a `chan struct{}` so you can keep it in a select if needed (not sure if it would be at the moment).

kolyshkin replied:
From where I stand, it looks like if a consumer is gone, there is no sense in trying to send a message we have. So, the current code:

```go
			// send the message unless the consumer is gone
￼ 			select {
			case watcher.Msg <- msg:
			case <-watcher.WatchConsumerGone():
￼ 				return
￼ 			}
```

is sufficient and should not be made any more complicated than that.

cpuguy83 replied:
For ConsumerCone, yes. Originally this was checking producer, I think?
I will need to do a full review of the entire file at this point to make sure I'm understanding all the implications of these changes.

kolyshkin replied:
Originally there was no distinction between the producer and the consumer. Both were calling `(LogWatcher *).Close()`, the first one was delivered, the second one was discarded. In case the first one was the Producer (meaning a container has stopped), the code we're discussing worked. In case the first one was the Consumer (meaning a log reader has exited), the code would block.

Now, with a distinction between the Producer and the Consumer, in this piece of code where we send a message to the Consumer, we are only interested if the Consumer is still there (to make sure we can send the message). In this place it does not make sense to check if the Producer is still there or not.

cpuguy83 commented on integration/container/logs_test.go:
```diff
@@ -33,3 +35,135 @@ func TestLogsFollowTailEmpty(t *testing.T) {
+}
+
+// Test for #37391
+func TestLogsFollowGoroutineLeak(t *testing.T) {
```
I feel like this can be a unit test... trying this out.

cpuguy83 replied:
Let me know what you think: https://gist.github.com/cpuguy83/e538793de18c762608358ee0eaddc197

cpuguy83 replied:
WDYT about this unit test?

kolyshkin replied:
Thanks, from the first glance it makes sense; need some time to test and incorporate it.

kolyshkin replied:
The only problem is this test can't be used before the commit that introduces Producer/Consumer and fixes the bug. Solved by taking the second test case and changing it to old-style `lw.Close()`.

kolyshkin commented:
I do not like the current solution either, and will try to get rid of `ProducerGone()`. There are two cases, it looks like both can be fine without being notified about container stop:

1. Regular `ContainerLogs()` (a.k.a. `docker logs`). In this case LogWatcher should read/send the logs until EOF and exit.

2. `ContainerLogs(opts.follow=true)` (a.k.a. `docker logs --follow`). In this case it does not matter if container is stopped -- LogWatcher should try forwarding the logs no matter what until the consumer is gone (if we will decide to fix https://github.com/moby/moby/issues/37630).

The above is the current theory, I am not yet sure if it will work.

cpuguy83 commented:
I would be up for an entire rewrite of this code, though getting the bug fixed first would be great.
`followLogs` is pretty gnarly.

kolyshkin commented:
> The above is the current theory, I am not yet sure if it will work.

Looks like it won't. I was hoping that logWatcher will receive "file removed" event upon container stop; this is not the case (the reason seems to be that since the file is opened it is not actually removed). With that, I am not sure how does "handleRename()" works (if it works at all), but that's another issue.

kolyshkin commented:
The idea of getting rid of ProducerGone() would work if not for lack of REMOVE event which is not happening since the file is currently opened. Alas, we have to keep it.

kolyshkin commented:
OK, kernel reports the change of link-count to a file via inotify then an opened file is removed -- the only problem is fsnotify is ignoring it. There's a PR to fix that https://github.com/fsnotify/fsnotify/pull/205 but it is abandoned.

Now, if fsnotify is amended we can probably do without `ProducerGone()` (assuming we also want to fix https://github.com/moby/moby/issues/37630).

kolyshkin commented:
> There's a PR to fix that fsnotify/fsnotify#205 but it is abandoned.

Filed https://github.com/fsnotify/fsnotify/pull/260; will try to simplify this PR using it.

kolyshkin commented:
> Filed fsnotify/fsnotify#260; will try to simplify this PR using it.

It seems like it's working! Replacing this PR with https://github.com/moby/moby/pull/37656

kolyshkin commented:
Patch updated and simplified a little bit. This is not as simple as https://github.com/moby/moby/pull/37656, but it maintains backward compatibility (i.e. `ContainerLogs(follow=true)` exits when a container is stopped, for which we need to have `[Watch]ProducerGone`).

cpuguy83 commented on daemon/logger/adapter.go:
```diff
@@ -93,21 +93,12 @@ func (a *pluginAdapterWithRead) ReadLogs(config ReadConfig) *LogWatcher {
-				case watcher.Err <- errors.Wrap(err, "error decoding log message"):
-				case <-watcher.WatchClose():
-				}
+				watcher.Err <- errors.Wrap(err, "error decoding log message")
```
This should be done on a select or else this can block.

kolyshkin replied:
I have to admit I don't understand it :(

Here's what I see:

1. Since this channel is made with `make chan(error, 1)` it is buffered, and the buffer size is 1. So it would block if we'll send more than 1 error before someone receives it. I do not see how can we send more than 1 error from here.
 
2. In the same function, about 20 lines above this, there's a blocking send to the same channel.

cpuguy83 replied:
If there's a buffer, this should be fine.
It is true that the expectation is if there is an error sent that it should return.
I have considered using this to send errors for data corruption and attempting a recovery (by e.g. scanning ahead to the next json object)... but that's not the current behavior.

cpuguy83 committed:
daemon/logger/loggerutils: add TestFollowLogsClose

This test case checks that followLogs() exits once the reader is gone.
Currently it does not (i.e. this test is supposed to fail) due to #37391.

[kolyshkin@: test case Brian Goff, changelog and all bugs are by me]
Source: https://gist.github.com/cpuguy83/e538793de18c762608358ee0eaddc197

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Ditched my integration test, added @cpuguy83's unit tests. Rebased, pushed.

kolyshkin commented:
ci failures

experimental: 
> 01:19:08.601 FAIL: docker_api_swarm_service_test.go:33: DockerSwarmSuite.TestAPIServiceUpdatePort
> 01:46:27.213 FAIL: docker_cli_swarm_test.go:1610: DockerSwarmSuite.TestSwarmPublishDuplicatePorts

janky:
> 01:51:30.480 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

power:
> 01:52:48.271 FAIL: docker_api_swarm_test.go:359: DockerSwarmSuite.TestAPISwarmRaftQuorum

All of the above looks unrelated

windows  and z (hmm do we need increased timeout here?):
> 00:07:48.896 --- FAIL: TestFollowLogsProducerGone (1.00s)
> 00:07:48.896 	logfile_test.go:188: timeout waiting for followLogs() to finish
> 00:07:48.896 FAIL

kolyshkin commented:
TestFollowLogsProducerGone fails on CI (but not on my machine)... 
> 00:09:10.758 --- FAIL: TestFollowLogsProducerGone (3.00s)
> 00:09:10.758 	logfile_test.go:188: timeout waiting for followLogs() to finish
> 00:09:10.758 FAIL

It seems we're counting extra messages wrong (getting `lw.Err()` is counted as a message while it shouldn't). Pushed a proposed fix, let's see.

kolyshkin commented:
WRT this code: https://github.com/moby/moby/blob/082da188a3aaf6abf38a68af073d1a36e829925b/daemon/logger/loggerutils/logfile.go#L553-L563

it was introduced by https://github.com/moby/moby/pull/27782 and can now be removed (details here: https://github.com/moby/moby/pull/27782#issuecomment-416794490)

Going to do that in order to simplify the code.

kolyshkin committed:
TestFollowLogsProducerGone: add

This should test that
 - all the messages produced are delivered (i.e. not lost)
 - followLogs() exits

Loosely based on the test having the same name by Brian Goff, see
https://gist.github.com/cpuguy83/e538793de18c762608358ee0eaddc197

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
OK, I had to rewrite `TestFollowLogsProducerGone` almost from scratch. Also, removed ctx usage from followLogs() as per https://github.com/moby/moby/pull/27782#issuecomment-416794490.

kolyshkin commented:
Windows CI failure in `DockerSuite.TestLogsAPIFollowEmptyOutput` (inability to remove the container because its log file was still opened by some other process) is apparently fixed by commit https://github.com/moby/moby/pull/37576/commits/b3992ad6e09acf00dfb2f887441ce251451a8949. The current theory is by commits modified timings so that the container exited faster than the thread poller.

kolyshkin commented:
Giving it some stress testing...

kolyshkin commented:
> Giving it some stress testing (30 runs of the above mentioned test case)...
...which it passed with dignity! Now, let's do the same without filenotify patches

kolyshkin commented:
OK, fixed the windows CI failure/issue by fixing bugs in pkg/filenotify/poller (see PR #37734). The commits from this PR is currently included in this one, for the sake of Windows CI. Once #37734 is merged I will rebase.

kolyshkin commented:
> Windows CI failure <...> is apparently fixed by commit b3992ad.

Nope, the real fix is apparently commit 8269ff0dc5c. Anyway, all three of them make some sense.

kolyshkin commented:
@cpuguy83 are you aware of any (unit or integration) test cases to check log rotation? As this PR touches that precious code that deals with rotated log files, I want to assess whether it broke something in that area...

kolyshkin commented:
ci failures

janky:
> 01:52:53.778 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

power:
> 01:48:13.817 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection
> 02:05:53.173 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

z:
> 01:52:20.860 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

* TestSwarmClusterRotateUnlockKey is https://github.com/moby/moby/issues/33041
* TestAPISwarmLeaderElection is https://github.com/moby/moby/issues/32673

both are flaky and unrelated to this PR

kolyshkin commented:
Speaking of log rotation, I was not sure I hadn't broken anything in that area, so I wrote a small test:

```sh
ID=$(docker run -d --log-opt=max-size=50K --log-opt=max-file=20 busybox sh -c 'i=0; while test $i -lt 10000; do echo $i; let i++; sleep 0; done')
docker logs --follow $ID | tee $ID.logs | wc -l
```

If it works, it should print 10000 (number of logs lines received).

It works on current master as well as with the code from this PR. I am not sure if I should make an integration test case out of it, but in any case this can be addressed in a followup.

PS note if I remove `sleep 0` the test fails on both branches -- the logs are rotated way to fast to follow. This also means that the above test case has some empirical values built into it, and might not work on a system different from my laptop.

kolyshkin commented:
On my laptop, the success of following rotated logs depends on the output rate, not the log rotation rate. I guess it is because most time is spent decoding json :-\

cpuguy83 reviewed (approved):
This looks good, minor comment on timeouts in the test.
LGTM

kolyshkin commented on daemon/logger/loggerutils/logfile_test.go:
```diff
@@ -74,3 +76,128 @@ func TestTailFiles(t *testing.T) {
+	lw.ConsumerGone()
+	select {
+	case <-followLogsDone:
+	case <-time.After(1 * time.Second):
```
addressed

kolyshkin commented:
Test timeouts increased; rebased.

cpuguy83 reviewed (approved):
LGTM

kolyshkin commented:
CI is green :dancing_men:

vdemeester reviewed (approved):
LGTM :tiger:

thaJeztah reviewed (approved):
SGTM, and all green; let's merge 👍
